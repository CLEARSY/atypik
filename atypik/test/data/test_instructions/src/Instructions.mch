/* Instructions
* Author: vdemeulenaere
* Creation date: 05/06/2023
*/
MACHINE
    Instructions
    
ABSTRACT_VARIABLES 
    var, var2, var3, var4, var5, 
    var6, var7, var8, var9, var10,
    var11, var12, var13, var14, fct
    
SETS
    ABS_SET
    
    
INVARIANT 
    var : INT &
    var2 : BOOL &   
    var3 : NAT &
    var4 : REAL &
    var5 : INTEGER & 
    var6 : INT &
    var7 : REAL &
    var8 : INTEGER &
    var9 : REAL &
    var10 : REAL &
    var11 : ABS_SET &
    var12 : ABS_SET &
    var13 : INT &
    var14 : INT &
    fct : POW(INT * BOOL * BOOL)
    
INITIALISATION
    BEGIN 
        // Assignments
        var, var2 := 0, FALSE ||
        fct(0, FALSE) := FALSE ||
        // Skip
        skip ||
        // Assertion
        ASSERT var = 1 THEN
            var3 := var
        END ||
        // Choice
        CHOICE 
            var4 := 3.14
        OR
            var4 := 4.2
        END ||
        // Conditional
        IF var4 = 3.14 THEN
            var5 := 1
        ELSE IF var4 = 4.2 THEN
                var5 := 2
            ELSE
                var5 := 3
            END
        END ||
        // Select 
        SELECT var /= var5 THEN
            var6 := var
        WHEN var = 1 THEN 
            var6 := var + 1
        WHEN var = 2 THEN
            var6 := var + 2
        ELSE 
            var6 := var - 1
        END ||
        // Case 
        CASE var6 OF
            EITHER 0 THEN
                var7 := 1.0
            OR 1 THEN
                var7 := 2.0
            ELSE
                var7 := 3.0
            END
        END ||
        // Any
        ANY x1, x2 WHERE 
            x1 : INT &
            x2 : NAT &
            x1 ** 2 + x2 ** 2 = 25
        THEN
            var8 := x1 + x2
        END ||
        // Let
        LET x1, x2 BE 
            x1 = var4 &
            x2 = var4 * 2.0
        IN 
            var9 := x1 + x2 ||
            var10 := x1 - x2
        END ||
        // Becomes in
        var11, var12 :: (ABS_SET * ABS_SET) ||
        // Becomes such that
        var13, var14 : (var13 : INT & var13 > 0 & var14 /= var13)
        
    END
    
    
END    
