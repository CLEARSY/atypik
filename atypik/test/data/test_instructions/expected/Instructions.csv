Expression	Identifier	Line	Column	Type
"0"	83	99	47	INTEGER
"(ABS_SET) *s (ABS_SET)"	80	97	34	(POW (PRODUCT ABS_SET ABS_SET))
"x1 = var4"	74	90	13	REAL
"x2"	73	89	17	REAL
"x1"	72	89	13	REAL
"25"	70	84	33	INTEGER
"((INT) *s (BOOL)) *s (BOOL)"	25	32	26	(POW (PRODUCT (PRODUCT INTEGER BOOL) BOOL))
"1.0"	56	73	25	REAL
"INTEGER"	21	22	12	(POW INTEGER)
"INT"	17	18	11	(POW INTEGER)
"2.0"	75	91	25	REAL
"3"	46	57	25	INTEGER
"NAT"	19	20	12	(POW INTEGER)
"2"	45	55	25	INTEGER
"var12"	13	11	12	ABS_SET
"2"	67	84	29	INTEGER
"ABS_SET"	0	14	5	(POW ABS_SET)
"var13"	14	11	19	INTEGER
"(var) +i (1)"	49	64	25	INTEGER
"REAL"	20	21	12	(POW REAL)
"var5"	6	9	28	INTEGER
"{}"	1	-1	-1	(POW ABS_SET)
"var11"	12	11	5	ABS_SET
"var9"	10	10	23	REAL
"2"	51	66	27	INTEGER
"var3"	4	9	16	INTEGER
"var"	2	9	5	INTEGER
"x2 = (var4) *r (2.0)"	77	91	13	REAL
"var2"	3	9	10	BOOL
"(x1) -r (x2)"	79	94	25	REAL
"0"	30	37	22	INTEGER
"var10"	11	27	5	REAL
"var8"	9	10	17	INTEGER
"FALSE"	31	37	25	BOOL
"(INT) *s (BOOL)"	22	32	19	(POW (PRODUCT INTEGER BOOL))
"BOOL"	18	19	12	(POW BOOL)
"fct"	16	11	33	(POW (PRODUCT (PRODUCT INTEGER BOOL) BOOL))
"var7"	8	10	11	REAL
"2"	65	84	19	INTEGER
"(0) |-> (FALSE)"	34	-1	-1	(PRODUCT INTEGER BOOL)
"var14"	15	11	26	INTEGER
"1"	47	63	20	INTEGER
"POW(((INT) *s (BOOL)) *s (BOOL))"	28	32	11	(POW (POW (PRODUCT (PRODUCT INTEGER BOOL) BOOL)))
"0"	32	38	13	INTEGER
"fct((0) |-> (FALSE))"	35	38	12	BOOL
"(x1) +r (x2)"	78	93	24	REAL
"FALSE"	38	38	26	BOOL
"(var4) *r (2.0)"	76	91	23	REAL
"1"	39	42	22	INTEGER
"(x2) **i (2)"	68	84	26	INTEGER
"1"	43	53	21	INTEGER
"4.2"	41	49	21	REAL
"3.14"	42	52	19	REAL
"var6"	7	10	5	INTEGER
"4.2"	44	54	24	REAL
"1"	48	64	27	INTEGER
"(var) -i (1)"	54	68	25	INTEGER
"2.0"	58	75	25	REAL
"2"	50	65	20	INTEGER
"(var) +i (2)"	52	66	25	INTEGER
"1"	53	68	27	INTEGER
"var4"	5	9	22	REAL
"0"	55	72	20	INTEGER
"3.14"	40	47	21	REAL
"1"	57	74	16	INTEGER
"FALSE"	33	38	16	BOOL
"3.0"	59	77	25	REAL
"x1"	63	81	13	INTEGER
"(x1) +i (x2)"	71	86	24	INTEGER
"((x1) **i (2)) +i ((x2) **i (2))"	69	84	21	INTEGER
"x2"	64	81	17	INTEGER
"(x1) **i (2)"	66	84	16	INTEGER
